tidy3d/plugins/microwave/custom_path_integrals.py

```diff
@@ -0,0 +1,163 @@
+"""Helper classes for performing custom path integrals with fields on the Yee grid"""
+
+from __future__ import annotations
+
+import pydantic.v1 as pd
+import numpy as np
+import xarray as xr
+
+from typing import Literal, Union
+
+from ...constants import MICROMETER
+from ...components.data.monitor_data import FieldData, FieldTimeData, ModeSolverData
+from ...components.data.data_array import (
+    ScalarFieldDataArray,
+    ScalarFieldTimeDataArray,
+    ScalarModeFieldDataArray,
+)
+from ...components.data.data_array import FreqDataArray, TimeDataArray, FreqModeDataArray
+from ...components.base import cached_property
+from ...components.types import Axis, ArrayFloat2D
+from ...exceptions import DataError, SetupError
+
+from .path_integrals import AbstractAxesRH
+
+MonitorDataTypes = Union[FieldData, FieldTimeData, ModeSolverData]
+EMScalarFieldType = Union[ScalarFieldDataArray, ScalarFieldTimeDataArray, ScalarModeFieldDataArray]
+IntegralResultTypes = Union[FreqDataArray, FreqModeDataArray, TimeDataArray]
+FieldParameter = Literal["E", "H"]
+
+
+class CustomPathIntegral2D(AbstractAxesRH):
+    """Class for defining a custom path integral defined as a curve on an axis-aligned plane."""
+
+    axis: Axis = pd.Field(
+        2, title="Axis", description="Specifies dimension of the planar axis (0,1,2) -> (x,y,z)."
+    )
+
+    position: float = pd.Field(
+        ...,
+        title="Position",
+        description="Position of the plane along the ``axis``.",
+    )
+
+    vertices: ArrayFloat2D = pd.Field(
```

Comment:
I think we have a `Vertices` type in types.py, which is used for PolySlab. would it be useful here? or maybe better to keep things separate?

Reply:
I copied the definition from Polyslab where `ArrayFloat2D` is used. However, maybe the fact that I expect vertices in a different order would be confusing. You'll notice that I ask users to supply vertices as (z,x). which is different from `PolySlab`. Do you think I should change this to follow `PolySlab`?

Reply:
I do actually think a consistent convention throughout would be nice... we decided to use the dimension "popping" convention long ago.. but it was a bit of an arbitrary choice that we're stuck with in a lot of places.

@momchil-flex thoughts?

Reply:
It was a simple fix to revert to the polyslab ordering of vertex coordinates.